updated dependencies, fixed typescript imports error esmoduleinterop,…

JustFly1984 opened the pull request:
… fixed performance, added eslint-plugin-react-perf, fixed eslint warnings and errors, fixed issues with mergeRefs

JustFly1984 commented on .eslintrc.json:
```diff
@@ -5,15 +5,16 @@
+    "plugin:react-perf/recommended",
+    "prettier",
+    "prettier/react",
+    "prettier/@typescript-eslint"
```
order does matter, prettier should be last

JustFly1984 commented on .eslintrc.json:
```diff
@@ -5,15 +5,16 @@
+    "prettier/@typescript-eslint"
   ],
-  "plugins": ["@typescript-eslint", "react", "prettier", "react-hooks", "import", "jest"],
+  "plugins": ["@typescript-eslint", "react", "react-perf", "react-hooks", "import", "jest", "prettier"],
```
same here - prettier should be last.

JustFly1984 commented on examples/.eslintrc.json:
```diff
@@ -0,0 +1,20 @@
+{
```
tried to extend CRA eslint, but it doesn't kicks in, and failing to find eslint-plugin-react-perf-rules, which in turn fails to run CRA with unknown rule comments in code.

JustFly1984 commented on examples/src/App.js:
```diff
@@ -1,4 +1,4 @@
-import React from 'react'
+import * as React from 'react'
```
correct imports for esmoduleinterop: false

JustFly1984 commented on package.json:
```diff
@@ -47,7 +47,7 @@
     "build": "rollup -c && yarn copy && rimraf dist/__tests__",
     "prepare": "npm run build",
-    "eslint": "eslint --fix {src,examples/src}/**/*.{js,ts,jsx,tsx}",
+    "eslint": "eslint --fix src/**/*.{ts,tsx} examples/src/**/*.js",
```
fixed linting, otherwise eslint is failing.

JustFly1984 commented on src/canvas.tsx:
```diff
@@ -584,18 +584,18 @@ export const useCanvas = (props: UseCanvasProps): DomEventHandlers => {
     } = state.current
     sharedState.current = props
-  }, [size, defaultCam])
+  }, [size, defaultCam]) // is defaultCam intentional?
```
here is possible missing hook dependency. please clarify intention

drcmda replied:
updates the shared state model when the camera changes, which triggers all consumers. it's correct

JustFly1984 commented on src/canvas.tsx:
```diff
@@ -605,15 +605,15 @@ export const useCanvas = (props: UseCanvasProps): DomEventHandlers => {
       gl.outputEncoding = THREE.sRGBEncoding
     }
-  }, [shadowMap, colorManagement])
+  }, [shadowMap, colorManagement]) // React Hook React.useLayoutEffect has a missing dependency: 'gl'
```
here is possible missing hook dependency. please clarify intention

drcmda replied:
you can add gl

JustFly1984 commented on src/canvas.tsx:
```diff
@@ -633,9 +633,9 @@ export const useCanvas = (props: UseCanvasProps): DomEventHandlers => {
       state
     )
-  }, [ready, children, sharedState.current])
+  }, [ready, children, sharedState.current]) // React Hook React.useLayoutEffect has missing dependencies: 'Canvas' and 'defaultScene'. Either include them or remove the dependency array. Mutable values like 'sharedState.current' aren't valid dependencies because mutating them doesn't re-render the component.
```
here is possible missing hook dependency. please clarify intention

drcmda replied:
you can add canvas and defaultscene, sharedState.current is specific, it can't just be sharedState

JustFly1984 replied:
yes, I understand that ref object can't be used to trigger hook, but changing ref.current does not trigger re-render by itself.

drcmda replied:
sharedstate changes when [size, defaultCam] change, then it triggers a render, but these two can also be added to the render part, just a little more fragile because if i decide to adapt when sharedstate changes, i now must adapt it down there, too. i would leave it for now. this will all go away with zustand.

JustFly1984 replied:
<img width="422" alt="Screen Shot 2020-11-22 at 3 38 07 PM" src="https://user-images.githubusercontent.com/2255993/99899030-c1bed800-2cd8-11eb-9294-3e459cf84ce4.png">

drcmda replied:
i know but it's obviously wrong. if we remove that r3f doesnt update useThree subscribers.

JustFly1984 replied:
I've commented out a warning, and added comment.

drcmda replied:
great. im glad all these special cases will be gone in v6 - i cant see that they are a little weird

JustFly1984 commented on src/declarations.d.ts:
```diff
@@ -1 +1 @@
-declare module 'react-merge-refs'
+// declare module 'react-merge-refs' // removed to fix typings issues with primitives refs loosing types. react-merge-refs is written in typescript, and has it's own types.
```
this was preventing 'react-merge-refs' from infer the types in primitive jsx ref prop

JustFly1984 commented on src/targets/shared/web/ResizeContainer.tsx:
```diff
@@ -71,34 +73,48 @@ const ResizeContainer = React.memo(function ResizeContainer(props: ResizeContain
   // Render the canvas into the dom
   return (
-    <div ref={mergeRefs([bind, containerRef])} style={{ ...defaultStyles, ...style }} {...events} {...restSpread}>
+    <div ref={mergeRefs([useMeasureRef, containerRef])} style={memoStyle} {...events} {...restSpread}>
```
here is type inference fixed for primitive ref prop

drcmda replied:
i dont understand this part:

```jsx
const memoStyle = React.useMemo(
    function memo() {
      return { ...defaultStyles, ...style }
    },
    [style]
  )
```

it returns a function for some reason?

drcmda replied:
but if you think it's necessary, then i prefer

const memoStyle = React.useMemo(() => ({ ...defaultStyles, ...style }), [style])

JustFly1984 replied:
for readability it is not great, but for performance it is better. less GC - faster perf.
I get used to named functions, better for debugging and performance audit. 
I use eslint-plugin-react-perf, and it highlights all common with passing new array/object to the component/primitive.

drcmda replied:
yeah but that's too much, just not my style - i find it uncomfortable. with all the usememo one liners in the code it would blow it up 20%. we can use the memo, but imo every line counts.

JustFly1984 replied:
ok, I will change all memo functions to arrow.

JustFly1984 replied:
changed. Also I found incorrect exports with React.memo().
The way it was - components had Anonymous name in React tree in react-dev-tools. Now all of them will have correct names.

JustFly1984 commented on tsconfig.json:
```diff
@@ -3,7 +3,7 @@
     "module": "commonjs",
     "moduleResolution": "node",
-    "esModuleInterop": true,
+    "esModuleInterop": false,
```
here I fixed typescript import errors for consumer projects.

JustFly1984 commented on src/canvas.tsx:
```diff
@@ -269,10 +269,10 @@ export const useCanvas = (props: UseCanvasProps): DomEventHandlers => {
     // Make viewport backwards compatible
     state.current.viewport = getCurrentViewport as ViewportData
-  }, [invalidateFrameloop, vr, concurrent, noEvents, ready, size, defaultCam, gl])
+  }, [invalidateFrameloop, vr, concurrent, noEvents, ready, size, defaultCam, gl, getCurrentViewport])
```
here I've fixed hook dependency

JustFly1984 commented on src/canvas.tsx:
```diff
@@ -298,15 +298,15 @@ export const useCanvas = (props: UseCanvasProps): DomEventHandlers => {
 
     if (ready) invalidate(state)
-  }, [defaultCam, gl, size, updateDefaultCamera])
+  }, [defaultCam, gl, size, updateDefaultCamera, getCurrentViewport, ready])
```
here I've fixed one more hook dependency

JustFly1984 commented on src/canvas.tsx:
```diff
@@ -605,25 +606,28 @@ export const useCanvas = (props: UseCanvasProps): DomEventHandlers => {
       gl.outputEncoding = THREE.sRGBEncoding
     }
-  }, [shadowMap, colorManagement])
+  }, [shadowMap, colorManagement, gl])
```
fixed hook

JustFly1984 commented:
I've reduced lines for all useMemo I found, and also cleaned up a bit and fixed md files in last commit.

Fixed planet example n last commit. Please review, I think it is ok to merge and release.

JustFly1984 commented:
@drcmda travis build has failed without the log. Can you please restart ci/cd?

JustFly1984 commented:
nevermind, I've pushed small fix, it should succeed now.

JustFly1984 commented:
I have fixed tests threshold in last commit, and tests are succedeng. Please merge and release new version.

drcmda commented:
@JustFly1984  looks like there are still some problems:

import * as Reconciler from 'react-reconciler'

./src/targets/svg → dist/svg.cjs.js...
(!) Cannot call a namespace ('Reconciler')
src/renderer.tsx: (483:17)
481: //    NoTimeout
482: //  >
483: const Renderer = Reconciler<any, any, THREE.Object3D, any, any, any, any, any, any, any, any, any>({
                      ^
484:   now,
485:   createInstance,

drcmda commented:
more problems:

./src/pages/Intro.js
  Line 47:9:  Definition for rule 'react-perf/jsx-no-new-object-as-prop' was not found 

you did // // eslint etc etc etc which didnt work when you do yarn eslint. so now the linter is happy but it doesnt run

i pushed it to master, you can take a look there.

JustFly1984 commented:
@drcmda I've created new PR with the same branch, and now there is merge conflicts. Can you please resolve? It is a fix for Reconciler. The issue with eslint - I did not setup correct extension of CRA eslint. Do you mind if I will update CRA to 4.x ?

JustFly1984 commented:
nevermind - I'm closing pr and making new one from master - I've pulled your changes. Great work!

drcmda commented:
cra 4 doesnt work with customize cra imo, it couldnt alias into src.

drcmda commented:
@JustFly1984 https://github.com/pmndrs/react-three-fiber/commit/417768fbe0f5fe2f0b20d4c2a6092c6d5bec6694

one of the dependencies caused an infinite loop. it's a good reminder that linting is not always right.
